skbio/alignment/_tabular_msa.py

```diff
@@ -570,3 +549,61 @@ def reindex(self, key=None, keys=None):
             keys_.flags.writeable = False
 
         self._keys = keys_
+
+    @experimental(as_of='0.4.0-dev')
+    def append(self, sequence):
+        """Append a sequence to the MSA.
+
+        Parameters
+        ----------
+        sequence : IUPACSequence
+            Sequence to be appended. Must match the dtype of the MSA and the
+            length of the second dimension of the MSA.
+
+        Raises
+        ------
+        TypeError
+            If the type of the sequence does not match the dtype of the MSA.
+        ValueError
+            If the length of the sequence does not match the length of the
+            second dimension of the MSA.
+
+        Notes
+        -----
+        The MSA is not automatically re-aligned when a sequence is appended.
+        Therefore, this operation is not necessarily meaningful on its own.
+
+        Examples
+        --------
+        >>> from skbio import DNA, TabularMSA
+        >>> msa = TabularMSA([DNA('ACG'), DNA('AC-')])
+        >>> len(msa)
+        2
+        >>> msa.append(DNA('AAA'))
+        >>> len(msa)
+        3
+
+        """
+        msa_is_empty = len(self) == 0
+        if msa_is_empty:
+            dtype = type(sequence)
+            if not issubclass(dtype, IUPACSequence):
+                raise TypeError(
+                    "`sequences` must contain scikit-bio sequence objects "
+                    "that have an alphabet, not type %r" % dtype.__name__)
+            self._dtype = dtype
+            self._shape = _Shape(sequence=1, position=len(sequence))
+            self._seqs = [sequence]
+        elif type(sequence) != self.dtype:
```

Comment:
Would `isinstance` work here? e.g., `elif not isinstance(sequence,
self.dtype)`

On Tue, Sep 15, 2015 at 12:19 PM, Jai Ram Rideout <notifications@github.com>
wrote:

> In skbio/alignment/_tabular_msa.py
> <https://github.com/biocore/scikit-bio/pull/1121#discussion_r39546135>:
>
> > +        >>> msa.append(DNA('AAA'))
> > +        >>> len(msa)
> > +        3
> > +
> > +        """
> > +        msa_is_empty = len(self) == 0
> > +        if msa_is_empty:
> > +            dtype = type(sequence)
> > +            if not issubclass(dtype, IUPACSequence):
> > +                raise TypeError(
> > +                    "`sequences` must contain scikit-bio sequence objects "
> > +                    "that have an alphabet, not type %r" % dtype.__name__)
> > +            self._dtype = dtype
> > +            self._shape = _Shape(sequence=1, position=len(sequence))
> > +            self._seqs = [sequence]
> > +        elif type(sequence) != self.dtype:
>
> This post <http://bugs.python.org/issue13387> on the Python issue tracker
> also indicates this is the correct way to check for an exact type. See also
> the docs for assertIsInstance here
> <https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertNotIsInstance>,
> which give an example of exact type checking (assertIs(type(obj), cls)).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1121/files#r39546135>.
>